File: clariphish/routes/tracking.js
// routes/tracking.js - Email Tracking (Opens, Clicks, Submissions)
const express = require('express');
const router = express.Router();
const db = require('../database/db');

// Track email open (1x1 pixel)
router.get('/open/:trackingId', async (req, res) => {
  try {
    const trackingId = req.params.trackingId;
    const ipAddress = req.ip || req.connection.remoteAddress;
    const userAgent = req.get('user-agent');

    // Update recipient record
    const result = await db.run(`
      UPDATE recipients 
      SET opened = 1, opened_at = CURRENT_TIMESTAMP, ip_address = ?, user_agent = ?
      WHERE tracking_id = ? AND opened = 0
    `, [ipAddress, userAgent, trackingId]);

    // Update campaign stats if this was first open
    if (result.changes > 0) {
      const recipient = await db.get(`
        SELECT campaign_id FROM recipients WHERE tracking_id = ?
      `, [trackingId]);

      if (recipient) {
        await db.run(`
          UPDATE campaigns 
          SET opened_count = opened_count + 1
          WHERE id = ?
        `, [recipient.campaign_id]);
      }
    }

    // Return 1x1 transparent pixel
    const pixel = Buffer.from(
      'R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7',
      'base64'
    );
    res.writeHead(200, {
      'Content-Type': 'image/gif',
      'Content-Length': pixel.length,
      'Cache-Control': 'no-cache, no-store, must-revalidate'
    });
    res.end(pixel);
  } catch (err) {
    console.error('Error tracking open:', err);
    res.status(500).send('Error');
  }
});

// Track link click
router.get('/click/:trackingId', async (req, res) => {
  try {
    const trackingId = req.params.trackingId;
    const ipAddress = req.ip || req.connection.remoteAddress;
    const userAgent = req.get('user-agent');

    // Get recipient and campaign info
    const recipient = await db.get(`
      SELECT r.*, c.landing_url 
      FROM recipients r
      JOIN campaigns c ON r.campaign_id = c.id
      WHERE r.tracking_id = ?
    `, [trackingId]);

    if (!recipient) {
      return res.status(404).send('Invalid tracking link');
    }

    // Update recipient record
    const result = await db.run(`
      UPDATE recipients 
      SET clicked = 1, clicked_at = CURRENT_TIMESTAMP, ip_address = ?, user_agent = ?
      WHERE tracking_id = ? AND clicked = 0
    `, [ipAddress, userAgent, trackingId]);

    // Update campaign stats if this was first click
    if (result.changes > 0) {
      await db.run(`
        UPDATE campaigns 
        SET clicked_count = clicked_count + 1
        WHERE id = ?
      `, [recipient.campaign_id]);
    }

    // Redirect to landing page with tracking ID
    const landingUrl = recipient.landing_url || process.env.DEFAULT_LANDING_URL;
    const separator = landingUrl.includes('?') ? '&' : '?';
    res.redirect(`${landingUrl}${separator}tid=${trackingId}`);
  } catch (err) {
    console.error('Error tracking click:', err);
    res.status(500).send('Error');
  }
});

// Track form submission
router.post('/submit/:trackingId', async (req, res) => {
  try {
    const trackingId = req.params.trackingId;
    const submittedData = JSON.stringify(req.body);
    const ipAddress = req.ip || req.connection.remoteAddress;
    const userAgent = req.get('user-agent');

    // Get recipient
    const recipient = await db.get(`
      SELECT campaign_id FROM recipients WHERE tracking_id = ?
    `, [trackingId]);

    if (!recipient) {
      return res.status(404).json({ error: 'Invalid tracking ID' });
    }

    // Update recipient record
    const result = await db.run(`
      UPDATE recipients 
      SET submitted = 1, submitted_at = CURRENT_TIMESTAMP, 
          submitted_data = ?, ip_address = ?, user_agent = ?
      WHERE tracking_id = ? AND submitted = 0
    `, [submittedData, ipAddress, userAgent, trackingId]);

    // Update campaign stats if this was first submission
    if (result.changes > 0) {
      await db.run(`
        UPDATE campaigns 
        SET submitted_count = submitted_count + 1
        WHERE id = ?
      `, [recipient.campaign_id]);
    }

    res.json({ 
      success: true, 
      message: 'Data recorded',
      redirect_url: process.env.TRAINING_URL || 'https://example.com/security-training'
    });
  } catch (err) {
    console.error('Error tracking submission:', err);
    res.status(500).json({ error: 'Failed to record submission' });
  }
});

// Get tracking data for a campaign
router.get('/campaign/:campaignId', async (req, res) => {
  try {
    const recipients = await db.query(`
      SELECT 
        email,
        sent,
        opened,
        clicked,
        submitted,
        sent_at,
        opened_at,
        clicked_at,
        submitted_at,
        ip_address,
        user_agent
      FROM recipients
      WHERE campaign_id = ?
      ORDER BY email
    `, [req.params.campaignId]);

    res.json(recipients);
  } catch (err) {
    console.error('Error fetching tracking data:', err);
    res.status(500).json({ error: 'Failed to fetch tracking data' });
  }
});

// Get detailed submission data
router.get('/submissions/:campaignId', async (req, res) => {
  try {
    const submissions = await db.query(`
      SELECT 
        email,
        submitted_at,
        submitted_data,
        ip_address,
        user_agent
      FROM recipients
      WHERE campaign_id = ? AND submitted = 1
      ORDER BY submitted_at DESC
    `, [req.params.campaignId]);

    // Parse submitted data
    submissions.forEach(sub => {
      try {
        sub.submitted_data = JSON.parse(sub.submitted_data);
      } catch (e) {
        sub.submitted_data = {};
      }
    });

    res.json(submissions);
  } catch (err) {
    console.error('Error fetching submissions:', err);
    res.status(500).json({ error: 'Failed to fetch submissions' });
  }
});

module.exports = router;

File: api/tracking.go
package api

import (
	"encoding/base64"
	"encoding/json"
	"log"
	"net/http"
	"time"

	"github.com/clariphish/clariphish/db"
)

// 1x1 transparent pixel
var trackingPixel, _ = base64.StdEncoding.DecodeString(
	"iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAYAAAAfFcSJAAAADUlEQVR42mNk+M9QDwADhgGAWjR9awAAAABJRU5ErkJggg==")

func TrackOpen(w http.ResponseWriter, r *http.Request) {
	rid := r.URL.Query().Get("rid")
	if rid == "" {
		http.Error(w, "Invalid request", http.StatusBadRequest)
		return
	}

	// Get campaign target info
	var campaignID, targetID int
	var email string
	err := db.DB.QueryRow(`
		SELECT ct.campaign_id, ct.target_id, t.email
		FROM campaign_targets ct
		JOIN targets t ON t.id = ct.target_id
		WHERE ct.rid = ?`, rid).Scan(&campaignID, &targetID, &email)

	if err == nil {
		// Check if already opened
		var count int
		db.DB.QueryRow(`
			SELECT COUNT(*) FROM events
			WHERE campaign_target_id = (SELECT id FROM campaign_targets WHERE rid = ?)
			AND message = 'Email Opened'`, rid).Scan(&count)

		if count == 0 {
			// Record the open event
			var ctID int
			db.DB.QueryRow("SELECT id FROM campaign_targets WHERE rid = ?", rid).Scan(&ctID)

			db.DB.Exec(`
				INSERT INTO events (campaign_id, campaign_target_id, email, time, message, details)
				VALUES (?, ?, ?, ?, 'Email Opened', '')`,
				campaignID, ctID, email, time.Now())

			// Update campaign target status
			db.DB.Exec("UPDATE campaign_targets SET status = 'opened' WHERE rid = ?", rid)
		}
	}

	// Return tracking pixel
	w.Header().Set("Content-Type", "image/png")
	w.Header().Set("Cache-Control", "no-cache, no-store, must-revalidate")
	w.Write(trackingPixel)
}

func TrackClick(w http.ResponseWriter, r *http.Request) {
	rid := r.URL.Query().Get("rid")
	log.Printf("TrackClick called with RID: %s", rid)
	if rid == "" {
		http.Redirect(w, r, "/", http.StatusFound)
		return
	}

	// Get campaign target info
	var campaignID, targetID, pageID int
	var email, redirectURL string
	err := db.DB.QueryRow(`
		SELECT ct.campaign_id, ct.target_id, t.email, c.page_id
		FROM campaign_targets ct
		JOIN targets t ON t.id = ct.target_id
		JOIN campaigns c ON c.id = ct.campaign_id
		WHERE ct.rid = ?`, rid).Scan(&campaignID, &targetID, &email, &pageID)

	if err == nil {
		log.Printf("TrackClick: Found campaign_id=%d, email=%s", campaignID, email)

		// Record the click event
		var ctID int
		db.DB.QueryRow("SELECT id FROM campaign_targets WHERE rid = ?", rid).Scan(&ctID)
		log.Printf("TrackClick: campaign_target_id=%d", ctID)

		result, err := db.DB.Exec(`
			INSERT INTO events (campaign_id, campaign_target_id, email, time, message, details)
			VALUES (?, ?, ?, ?, 'Clicked Link', '')`,
			campaignID, ctID, email, time.Now())

		if err != nil {
			log.Printf("TrackClick: Error inserting event: %v", err)
		} else {
			eventID, _ := result.LastInsertId()
			log.Printf("TrackClick: Event created with ID=%d", eventID)
		}

		// Update campaign target status
		db.DB.Exec("UPDATE campaign_targets SET status = 'clicked' WHERE rid = ?", rid)

		// Get page HTML
		var pageHTML string
		db.DB.QueryRow("SELECT html FROM pages WHERE id = ?", pageID).Scan(&pageHTML)

		// Inject tracking form if capture is enabled
		db.DB.QueryRow("SELECT redirect_url FROM pages WHERE id = ?", pageID).Scan(&redirectURL)

		// Return the landing page with RID embedded
		w.Header().Set("Content-Type", "text/html")
		w.Write([]byte(injectRID(pageHTML, rid)))
		return
	} else {
		log.Printf("TrackClick: Error finding campaign target: %v", err)
	}

	http.Redirect(w, r, "/", http.StatusFound)
}

func TrackSubmission(w http.ResponseWriter, r *http.Request) {
	rid := r.FormValue("rid")
	log.Printf("TrackSubmission called with RID: %s", rid)
	if rid == "" {
		http.Error(w, "Invalid request", http.StatusBadRequest)
		return
	}

	// Parse form data
	r.ParseForm()
	formData := make(map[string]interface{})
	for key, values := range r.Form {
		if key != "rid" && len(values) > 0 {
			formData[key] = values[0]
		}
	}
	log.Printf("TrackSubmission: Form data: %+v", formData)

	// Get campaign target info
	var campaignID int
	var email string
	err := db.DB.QueryRow(`
		SELECT ct.campaign_id, t.email
		FROM campaign_targets ct
		JOIN targets t ON t.id = ct.target_id
		WHERE ct.rid = ?`, rid).Scan(&campaignID, &email)

	if err == nil {
		log.Printf("TrackSubmission: Found campaign_id=%d, email=%s", campaignID, email)

		// Record the submission event
		var ctID int
		db.DB.QueryRow("SELECT id FROM campaign_targets WHERE rid = ?", rid).Scan(&ctID)
		log.Printf("TrackSubmission: campaign_target_id=%d", ctID)

		detailsJSON, _ := json.Marshal(formData)
		result, err := db.DB.Exec(`
			INSERT INTO events (campaign_id, campaign_target_id, email, time, message, details)
			VALUES (?, ?, ?, ?, 'Submitted Data', ?)`,
			campaignID, ctID, email, time.Now(), string(detailsJSON))

		if err != nil {
			log.Printf("TrackSubmission: Error inserting event: %v", err)
		} else {
			eventID, _ := result.LastInsertId()
			log.Printf("TrackSubmission: Event created with ID=%d, details=%s", eventID, string(detailsJSON))
		}

		// Update campaign target status
		db.DB.Exec("UPDATE campaign_targets SET status = 'submitted' WHERE rid = ?", rid)

		// Get redirect URL
		var redirectURL string
		db.DB.QueryRow(`
			SELECT p.redirect_url FROM pages p
			JOIN campaigns c ON c.page_id = p.id
			WHERE c.id = ?`, campaignID).Scan(&redirectURL)

		if redirectURL != "" {
			respondJSON(w, map[string]string{"redirect": redirectURL})
			return
		}
	} else {
		log.Printf("TrackSubmission: Error finding campaign target: %v", err)
	}

	respondJSON(w, map[string]bool{"success": true})
}

func injectRID(html, rid string) string {
	// Inject RID into forms
	injectionScript := `<script>
document.addEventListener('DOMContentLoaded', function() {
	var forms = document.getElementsByTagName('form');
	for (var i = 0; i < forms.length; i++) {
		var ridInput = document.createElement('input');
		ridInput.type = 'hidden';
		ridInput.name = 'rid';
		ridInput.value = '` + rid + `';
		forms[i].appendChild(ridInput);

		forms[i].addEventListener('submit', function(e) {
			e.preventDefault();
			var formData = new FormData(this);
			fetch('/api/report', {
				method: 'POST',
				body: formData
			}).then(function(response) {
				return response.json();
			}).then(function(data) {
				if (data.redirect) {
					window.location.href = data.redirect;
				} else {
					alert('Thank you for your submission!');
				}
			});
		});
	}
});
</script></body>`

	return replaceOnce(html, "</body>", injectionScript)
}

func replaceOnce(s, old, new string) string {
	i := len(s) - len(old)
	for i >= 0 {
		if s[i:i+len(old)] == old {
			return s[:i] + new + s[i+len(old):]
		}
		i--
	}
	return s + new
}

// TrackReportPhishing tracks when a user reports the email as phishing
func TrackReportPhishing(w http.ResponseWriter, r *http.Request) {
	rid := r.URL.Query().Get("rid")
	log.Printf("TrackReportPhishing called with RID: %s", rid)
	if rid == "" {
		http.Error(w, "Invalid request", http.StatusBadRequest)
		return
	}

	// Get campaign target info
	var campaignID int
	var email string
	err := db.DB.QueryRow(`
		SELECT ct.campaign_id, t.email
		FROM campaign_targets ct
		JOIN targets t ON t.id = ct.target_id
		WHERE ct.rid = ?`, rid).Scan(&campaignID, &email)

	if err == nil {
		log.Printf("TrackReportPhishing: Found campaign_id=%d, email=%s", campaignID, email)

		// Check if already reported
		var count int
		db.DB.QueryRow(`
			SELECT COUNT(*) FROM events
			WHERE campaign_target_id = (SELECT id FROM campaign_targets WHERE rid = ?)
			AND message = 'Reported Phishing'`, rid).Scan(&count)

		log.Printf("TrackReportPhishing: Existing report count=%d", count)

		if count == 0 {
			// Record the report event
			var ctID int
			db.DB.QueryRow("SELECT id FROM campaign_targets WHERE rid = ?", rid).Scan(&ctID)
			log.Printf("TrackReportPhishing: campaign_target_id=%d", ctID)

			result, err := db.DB.Exec(`
				INSERT INTO events (campaign_id, campaign_target_id, email, time, message, details)
				VALUES (?, ?, ?, ?, 'Reported Phishing', '')`,
				campaignID, ctID, email, time.Now())

			if err != nil {
				log.Printf("TrackReportPhishing: Error inserting event: %v", err)
			} else {
				eventID, _ := result.LastInsertId()
				log.Printf("TrackReportPhishing: Event created with ID=%d", eventID)
			}

			// Update campaign target status
			db.DB.Exec("UPDATE campaign_targets SET status = 'reported' WHERE rid = ?", rid)
			log.Printf("TrackReportPhishing: Updated status to 'reported'")
		} else {
			log.Printf("TrackReportPhishing: Already reported, skipping")
		}
	} else {
		log.Printf("TrackReportPhishing: Error finding campaign target: %v", err)
	}

	// Return a user-friendly HTML page
	w.Header().Set("Content-Type", "text/html")
	w.Write([]byte(`
<!DOCTYPE html>
<html>
<head>
	<title>Thank You for Reporting</title>
	<style>
		body { font-family: Arial, sans-serif; text-align: center; padding: 50px; background: #f5f5f5; }
		.container { background: white; padding: 40px; border-radius: 8px; max-width: 500px; margin: 0 auto; box-shadow: 0 2px 4px rgba(0,0,0,0.1); }
		h1 { color: #28a745; }
		p { color: #666; line-height: 1.6; }
	</style>
</head>
<body>
	<div class="container">
		<h1>✓ Thank You!</h1>
		<p>Your report has been recorded successfully.</p>
		<p>Thank you for being vigilant about suspicious emails and helping us improve security awareness.</p>
	</div>
</body>
</html>
	`))
}
